### PR TITLE
Switch from launch configurations to launch templates

### LINKED ISSUE #39: Check if the installer uses Auto Scaling launch configurations
check if in the Terraform module we use under the hood launch templates.
because launch configuration are going to be dismissed december 1

### PR DESCRIPTION
**Pull request by lnovara**
Given the upcoming [deprecation of launch configuration](https://aws.amazon.com/blogs/compute/amazon-ec2-auto-scaling-will-no-longer-add-support-for-new-ec2-features-to-launch-configurations/), we need to migrate to launch templates. As you can see, the [underlying Terraform module](https://github.com/terraform-aws-modules/terraform-aws-eks) already supports launch templates and the changes needed to the installer are trivial but making the switch for existing clusters would imply downtime as all the nodes need to be destroyed and then recreated.

How do we want to manage the migration? I think the heavy lifting should be done by furyctl (I'm trying to figure how) but we need an installer that can support but launch configuration and launch templates for the sake of migration.

What are your thoughts on the topic?

Closes #39

### CODE REVIEW

**Comment by nutellinoit** [comment]
Maybe a conditional dynamic block with a parameter that pilot if we are using `worker_groups` or `worker_groups_launch_template` ?

https://codeinthehole.com/tips/conditional-nested-blocks-in-terraform/

**Comment by lnovara** [comment]
In the end, dynamic blocks where not needed. I managed to achieve something meaningful using just conditionals. Obviously, in the installer, it's up to the user to select the value and apply the correct steps to migrate from launch configurations to launch templates.

**Comment by ralgozino** [comment]
Hi @lnovara 
From the announcement you linked, it seems that the Auto Scaling Groups can be switched from launch configuration to launch templates without needing to recreate them.

Maybe the only thing to have in mind is the order that the operations are performed (first add the launch template field and then remove the launch configuration)

I wonder if changing the value makes Terraform recreate the ASG or if it is smart enough, did you test this by any chance?

Thanks!

**Review by ralgozino — CHANGES_REQUESTED** [review]
After reviewing together the missing bits of the PR are:

- rename `node_pool_kind` to `node_pool_launch_kind`
- add a markdown file with the procedure to follow to migrate from one kind to the other

**Review by omissis — APPROVED** [review]

**Comment by lnovara** [comment]
We also tried to make the transition as transparent as possible to the end user and without any downtime, but that migration path involved doing nasty stuff such as manually editing the Terraform state so, even if we had some success, we gave up in that route.

File: docs/UPGRADE-1.11.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Upgrade from v1.10.x to v1.11.x
```
Comment (ralgozino) [inline docs/UPGRADE-1.11.md:1]: 1.10 was a draft and has not been released yet, This should read 

```markdown
# Upgrade from v1.9.x to v1.10.x
```

File: docs/UPGRADE-1.11.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+In this version of `fury-eks-installer`, we had to introduce support for launch templates due to the [deprecation of launch configurations](https://aws.amazon.com/blogs/compute/amazon-ec2-auto-scaling-will-no-longer-add-support-for-new-ec2-features-to-launch-configurations/).
+
+To achieve this goal, we introduced a new variable `node_pools_launch_kind` (that defaults to `launch_templates`) to select wheter to use launch templates, launch configurations or both.
+
```
Comment (ralgozino) [inline docs/UPGRADE-1.11.md:6]: For some reason I can't make code suggestions 🤷 
I'll add this bit here:

```markdown
- Selecting `launch_configuration` in existing Fury clusters would change nothing.
- Selecting `launch_template` for existing clusters will delete the current node pools and create new ones.
- Selecting `both` for existing clusters will create new node pools that use the `launch_templates` and leave the old with `launch_configuration` intact.

Continue reading for how to migrate an existing cluster from `launch_configuration` to `launch_templates`.

```

File: docs/UPGRADE-1.11.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+## Migrate from launch configurations to launch templates
+
+1. Cordon all the existing nodes created using launch configurations using `kubectl cordoni <node_name>`.
```
Comment (ralgozino) [inline docs/UPGRADE-1.11.md:9]: there's a typo in the command `cordoni` instead of `cordon` 😄 🇮🇹

File: docs/UPGRADE-1.11.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+> ⚠️ **WARNING**
+> If any of the node fails before migrating to launch templates, the pods will have nowhere to be scheduled. If you can't cordon all the nodes at once, take note of the existing nodes and start cordoning them after the new nodes from the launch templates start joining the cluster.
+
+2. Add `node_pools_launch_kind = "both"` to your Terraform module configuration and apply.
+
+3. Wait for the new nodes to join the cluster.
+
+4. Drain the old nodes that you cordoned in step 1 using `kubectl drain --ignore-daemonsets --delete-local-data <node_name>`. Now all the pods are running on nodes created with launch templates.
+
+5. Change `node_pools_launch_kind` to `"launch_templates"` in your Terraform module configuration and apply.
```
Comment (ralgozino) [inline docs/UPGRADE-1.11.md:20]: I'd add a comment here saying that this will effectively delete the old nodes.

**Review by ralgozino — CHANGES_REQUESTED** [review]

**Review by ralgozino — APPROVED** [review]
:shipit:

**Pull request merged**